File: java/src/main/be/steinsomers/bron_kerbosch/BronKerbosch3_MT.java
package be.steinsomers.bron_kerbosch;

import lombok.NoArgsConstructor;
import lombok.RequiredArgsConstructor;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public final class BronKerbosch3_MT implements BronKerboschAlgorithm {
    @Override
    public void explore(final UndirectedGraph graph, final Consumer<int[]> cliqueConsumer) throws InterruptedException {
        final var worker = new Worker(graph, cliqueConsumer);
        worker.work();
    }

    @RequiredArgsConstructor
    private static class StartJob {
        private static final int CLEAN_END_VERTEX = -1;
        private static final int DIRTY_END_VERTEX = -2;
        final int startVertex;

        static StartJob CleanEnd() {
            return new StartJob(CLEAN_END_VERTEX);
        }

        static StartJob DirtyEnd() {
            return new StartJob(DIRTY_END_VERTEX);
        }

        final boolean isGenuine() {
            return startVertex >= 0;
        }
    }

    private static final class VisitJob extends StartJob {
        VisitJob(final StartJob start, final Set<Integer> candidates, final Set<Integer> excluded) {
            super(start.startVertex);
            this.candidates = candidates;
            this.excluded = excluded;
        }

        final Set<Integer> candidates;
        final Set<Integer> excluded;
    }

    private static final class Worker {
        private static final int NUM_VISITING_THREADS = 5;

        private final UndirectedGraph graph;
        private final Consumer<int[]> cliqueConsumer;
        private final BlockingQueue<StartJob> startQueue;
        private final BlockingQueue<VisitJob> visitQueue;
        private final StartProducer startProducer = new StartProducer();
        private final VisitProducer visitProducer = new VisitProducer();

        @NoArgsConstructor
        private final class StartProducer implements Runnable {
            @Override
            public void run() {
                try {
                    final Iterable<Integer> vertices = () -> new DegeneracyOrdering(graph, -1);
                    for (final var v : vertices) {
                        startQueue.put(new StartJob(v));
                    }
                    startQueue.put(StartJob.CleanEnd());
                } catch (final InterruptedException consumed) {
                    startQueue.clear();
                    startQueue.add(StartJob.DirtyEnd());
                }
            }
        }

        @NoArgsConstructor
        private final class VisitProducer implements Runnable {
            @SuppressWarnings("UseOfConcreteClass")
            @Override
            public void run() {
                try {
                    final Set<Integer> mut_excluded = new HashSet<>(graph.order());
                    StartJob startJob;
                    while ((startJob = startQueue.take()).isGenuine()) {
                        final var v = startJob.startVertex;
                        final var neighbours = graph.neighbours(v);
                        assert !neighbours.isEmpty();
                        final var candidates = util.Difference(neighbours, mut_excluded)
                                .collect(Collectors.toCollection(HashSet::new));
                        if (candidates.isEmpty()) {
                            assert !util.AreDisjoint(neighbours, mut_excluded);
                        } else {
                            final var excluded = util.Intersect(neighbours, mut_excluded)
                                    .collect(Collectors.toCollection(HashSet::new));
                            final VisitJob job = new VisitJob(startJob, candidates, excluded);
                            visitQueue.put(job);
                        }
                        mut_excluded.add(v);
                    }
                    for (int i = 0; i < NUM_VISITING_THREADS; ++i) {
                        visitQueue.put(new VisitJob(startJob, null, null));
                    }
                } catch (InterruptedException _) {
                    visitQueue.clear();
                    for (int i = 0; i < NUM_VISITING_THREADS; ++i) {
                        visitQueue.add(new VisitJob(StartJob.DirtyEnd(), null, null));
                    }
                }
            }
        }

        @NoArgsConstructor
        private final class Visitor implements Runnable {
            @SuppressWarnings("UseOfConcreteClass")
            @Override
            public void run() {
                try {
                    VisitJob job;
                    while ((job = visitQueue.take()).isGenuine()) {
                        BronKerboschPivot.visit(graph, cliqueConsumer,
                                PivotChoice.MaxDegreeLocal,
                                job.candidates,
                                job.excluded,
                                new int[]{job.startVertex});
                    }
                } catch (final InterruptedException consumed) {
                    // assume that thread remembers it
                }
            }
        }

        private Worker(final UndirectedGraph graph, final Consumer<int[]> cliqueConsumer) {
            this.graph = graph;
            this.cliqueConsumer = cliqueConsumer;
            startQueue = new ArrayBlockingQueue<>(64);
            visitQueue = new ArrayBlockingQueue<>(64);
        }

        void work() throws InterruptedException {
            new Thread(startProducer).start();
            new Thread(visitProducer).start();
            final var visitors = new Thread[NUM_VISITING_THREADS];
            for (int i = 0; i < NUM_VISITING_THREADS; ++i) {
                visitors[i] = new Thread(new Visitor());
                visitors[i].start();
            }
            for (int i = 0; i < NUM_VISITING_THREADS; ++i) {
                visitors[i].join();
            }
            for (int i = 0; i < NUM_VISITING_THREADS; ++i) {
                if (visitors[i].isInterrupted()) {
                    throw new InterruptedException("Some thread got interrupted");
                }
            }
        }
    }
}

File: kotlin/src/main/be/steinsomers/bron_kerbosch/BronKerbosch3MT.kt
package be.steinsomers.bron_kerbosch

import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.BlockingQueue

class BronKerbosch3MT : BronKerboschAlgorithm {
    companion object {
        private const val NUM_VISITING_THREADS = 5
    }

    @Throws(InterruptedException::class)
    override fun explore(graph: UndirectedGraph, cliqueConsumer: CliqueConsumer) {
        val worker = Worker(graph, cliqueConsumer)
        worker.work()
    }

    private sealed class VisitJob {
        object CleanEnd : VisitJob()
        object DirtyEnd : VisitJob()
        data class Work(
            val startVertex: Vertex, val candidates: MutableSet<Vertex>, val excluded: MutableSet<Vertex>
        ) : VisitJob() {
            init {
                require(startVertex.index >= 0) // as if it would enable Kotlin to enumerate the end cases as negatives
            }
        }
    }

    private class Worker(private val graph: UndirectedGraph, private val cliqueConsumer: CliqueConsumer) {
        private val degeneracy = GraphDegeneracy(graph)
        private val visitQueue: BlockingQueue<VisitJob> = ArrayBlockingQueue(64)

        private inner class VisitProducer : Runnable {
            override fun run() {
                try {
                    degeneracy.forEach { v: Vertex ->
                        val (neighbouringCandidates, neighbouringExcluded) =
                            Util.partition(graph.neighbours(v)) { v -> degeneracy.isCandidate(v) }
                        Debug.assert { neighbouringCandidates.isNotEmpty() }
                        visitQueue.put(VisitJob.Work(v, neighbouringCandidates, neighbouringExcluded))
                    }
                    repeat(NUM_VISITING_THREADS) { _ -> visitQueue.put(VisitJob.CleanEnd) }
                } catch (_: InterruptedException) {
                    repeat(NUM_VISITING_THREADS) { _ -> visitQueue.put(VisitJob.DirtyEnd) }
                }
            }
        }

        private inner class Visitor : Runnable {
            override fun run() {
                while (true) {
                    when (val job = visitQueue.take()) {
                        is VisitJob.CleanEnd -> return
                        is VisitJob.DirtyEnd -> return
                        is VisitJob.Work ->
                            BronKerboschPivot.visit(
                                graph = graph, cliqueConsumer = cliqueConsumer,
                                pivotChoice = PivotChoice.MaxDegreeLocal,
                                candidates = job.candidates,
                                excluded = job.excluded,
                                cliqueInProgress = CliqueInProgress.singleton(job.startVertex)
                            )
                    }
                }
            }
        }

        @Throws(InterruptedException::class)
        fun work() {
            val visitorProducer = Thread(VisitProducer())
            val visitors = Array(NUM_VISITING_THREADS) { Thread(Visitor()) }
            visitorProducer.start()
            visitors.forEach { v -> v.start() }
            visitors.forEach { v -> v.join() }
        }
    }
}
